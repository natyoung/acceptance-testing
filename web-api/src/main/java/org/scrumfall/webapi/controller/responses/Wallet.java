package org.scrumfall.webapi.controller.responses;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

@EqualsAndHashCode
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE, force = true)
public class Wallet {
    private final String id;
    private final String userId;
    private final int balance;

    public Wallet(String id, String userId, int balance) {
        this.id = id;
        this.userId = userId;
        this.balance = balance;
    }
}
