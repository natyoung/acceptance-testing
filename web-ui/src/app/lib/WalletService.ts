import {Wallet} from "@/app/lib/definitions";
import axios, {AxiosResponse} from "axios";

export default class WalletService {
  private host: string;

  constructor(host: string) {
    this.host = host;
  }

  async getWallet(userId: string): Promise<AxiosResponse> {
    return axios.get<Wallet>(`${this.host}/wallets/${userId}`)
      .then(res => res)
  }

  async addFunds(userId: string, amount: number): Promise<AxiosResponse> {
    return axios.patch<Wallet>(`${this.host}/wallets/${userId}`, {
      amount: amount
    })
      .then(res => res)
  }
}
