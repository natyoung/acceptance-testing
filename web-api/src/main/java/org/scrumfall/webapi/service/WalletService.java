package org.scrumfall.webapi.service;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import lombok.Setter;
import org.scrumfall.webapi.controller.responses.Wallet;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

import static org.springframework.util.MimeTypeUtils.APPLICATION_JSON_VALUE;

@Service
public class WalletService {
    private static final Gson GSON = new Gson();

    private final HttpClient httpClient;

    @Setter
    private String host = System.getenv("HOST_WALLET");

    @Autowired
    public WalletService(HttpClient httpClient) {
        this.httpClient = httpClient;
    }

    public Optional<Wallet> getByUserId(String id) {
        try {
            final URI uri = new URI(host + "/wallets/" + URLEncoder.encode(id, StandardCharsets.UTF_8));
            final HttpRequest request = HttpRequest.newBuilder()
                    .uri(uri)
                    .GET()
                    .header(HttpHeaders.CONTENT_TYPE, APPLICATION_JSON_VALUE)
                    .build();
            final HttpResponse<String> response = httpClient
                    .send(request, HttpResponse.BodyHandlers.ofString());
            final Wallet wallet = GSON.fromJson(response.body(), Wallet.class);
            return Optional.ofNullable(wallet);
        } catch (URISyntaxException | IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<Wallet> pay(String id, int amount) {
        try {
            final URI uri = new URI(host + "/wallets/" + URLEncoder.encode(id, StandardCharsets.UTF_8));

            final JsonObject requestBody = new JsonObject();
            requestBody.addProperty("amount", amount);

            final HttpRequest request = HttpRequest.newBuilder()
                    .uri(uri)
                    .method("PATCH", HttpRequest.BodyPublishers.ofString(requestBody.toString()))
                    .header(HttpHeaders.CONTENT_TYPE, APPLICATION_JSON_VALUE)
                    .build();
            final HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            final Wallet wallet = GSON.fromJson(response.body(), Wallet.class);
            return Optional.ofNullable(wallet);
        } catch (URISyntaxException | IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}
