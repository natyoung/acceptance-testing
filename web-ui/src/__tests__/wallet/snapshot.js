import {render, waitFor} from '@testing-library/react'
import Wallet from '../../app/wallet/page'
import WalletService from '@/app/lib/WalletService'
jest.mock('next/image')

beforeEach(() => {
  jest.spyOn(WalletService.prototype, 'getWallet')
    .mockImplementation(() => Promise.resolve(
      {id: '', balance: 1}
    ));
});

it('renders wallet page unchanged', async () => {
  const {container} = await waitFor(() => {
    return render(<Wallet/>);
  })

  expect(container).toMatchSnapshot()
})
