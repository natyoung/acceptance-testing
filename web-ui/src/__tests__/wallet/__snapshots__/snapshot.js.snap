// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders wallet page unchanged 1`] = `
<div>
  <main
    class="flex min-h-screen flex-col items-center justify-between"
  >
    <div
      class="heroWrapper"
    >
      <div
        class="imageWrapper"
      />
    </div>
    <div
      class="mb-32 grid text-center lg:max-w-10xl lg:w-full lg:mb-0 lg:grid-cols-1 lg:text-center"
    >
      <h1
        class="text-4xl"
      >
        Wallet
      </h1>
      <form
        class="bg-dark shadow-md rounded px-8 pt-6 pb-8 mb-4"
      >
        <div
          class="mb-4"
        >
          <label
            class="block text-blue-500  text-sm font-bold mb-2"
            for="amount"
          >
            Story Points delivered to the dev environment:
          </label>
          <input
            class="shadow appearance-none border rounded py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            data-testid="amount"
            name="amount"
            type="text"
          />
        </div>
        <div
          class="items-center justify-between mb-4"
        >
          <button
            class="bg-blue-800 hover:bg-blue-600 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
            data-testid="submit-amount"
            type="submit"
          >
            Add Points
          </button>
        </div>
      </form>
    </div>
  </main>
</div>
`;
